pdfminer/converter.py

```diff
@@ -83,14 +83,13 @@ def paint_path(self, gstate, stroke, fill, evenodd, path):
                 self.paint_path(gstate, stroke, fill, evenodd, subpath)
 
         else:
-            if shape == 'ml':
+            if shape == 'mlh':
```

Comment:
This `shape == 'ml'` check has been in pdfminer since 2010. Do you think it has never matched before?

Reply:
Interesting! It seems possible. Going back to the oldest tag in the repo — `20160614` — this short script:

```python
from pdfminer.converter import PDFLayoutAnalyzer
from pdfminer.layout import LTContainer
from pdfminer.pdfinterp import PDFGraphicState

analyzer = PDFLayoutAnalyzer(None)
analyzer.set_ctm([1, 0, 0, 1, 0, 0])
analyzer.cur_item = LTContainer([0, 1000, 0, 1000])

path = [
    ("m", 10, 30),
    ("l", 10, 40),
    ("h",),
]

analyzer.paint_path(PDFGraphicState(), False, False, False, path)
objs = analyzer.cur_item._objs

print(objs) 
```

... outputs `[<LTCurve 10.000,30.000,10.000,40.000>]` (instead of `<LTLine ...>`).

But perhaps there was earlier code that meant the path passed to `.paint_path` would not have been this:

```python
[
    ("m", 10, 30),
    ("l", 10, 40),
    ("h",),
]
```

... but rather this:

```python
[
    ("m", 10, 30),
    ("l", 10, 40),
]
```

... in which case it would have matched?

Reply:
The pdf reference says:

> Close  the  current  subpath  by  appending  a  straight  line  segment  from  the  current  point  to  the  starting  point  of  the  subpath.  If  the  current subpath is already closed, h does nothing. 
> This  operator  terminates  the  current  subpath.  Appending  another  segment to the current path begins a new subpath, even if the new segment begins at the endpoint reached by the h operation. 

So in proper pdf's `h` should be used. I did a quick search on all the pdf`s in our test set, printing all the `path` variables, and none of them has a `ml` without `h` path. 

So I think we should proceed with matching the implementation to the pdf reference manual. In any case we have the `LTCurve` as a backup.

Reply:
Thanks, and very interesting. Inspired by your search, I have spotted an `h`-less line in one of the PDFs I parse regularly. And it seems PDF-rendering software accepts/renders respects those (i.e., they are rendered). So perhaps it makes most sense to capture both possibilities by tweaking to `if shape in ('mlh', 'ml'):`?

And perhaps it's worth considering something similar for the rect-cactcher, i.e., handling `mllll` shapes where the final point is the same as the initial ... or perhaps that is too rare/complex an edge-case?

Reply:
Does the `h`-less PDF you found give an error with the current code from this MR? Then lets use the `if shape in {'mlh', 'ml'}:` approach. And add the pdf to the contrib test cases to make sure we keep supporting this. Also make sure to document why this pdf was added. And to add a comment to `'ml'` about it being an anomaly that we want to support. 

Its tempting to make the rect-catcher also more generic, but I prefer to implement generic stuff when we have an example PDF that show that its happening.

Reply:
My apologies for the delay in responding, and many thanks for your thoughts here. In examining a small corpus of PDFs, I have found that the `ml` variation is somewhat common — common enough to support it, I think. 

Accordingly, I have made the changes I believe you have suggested — In addition to the main code adjustment, I have added a comment noting the `ml` variation, as well as a couple of new test assertions, including one that uses a new `samples/contrib` PDF. (It's one page from the United States' Federal Register publication, so it is in the public domain.)